public/static/js/init.py

```diff
@@ -514,10 +515,53 @@ def pyexec(code, expected_input, expected_output):
         criteria_outcomes = []
         for requirement in expected_output:
             requirement = requirement.as_object_map()
-            pattern = requirement["pattern"]  # the pattern to match. Must be present
             typ = requirement.get("typ", "+")
-
-            test_string = test_output.buffer if typ[0] != "c" else code
+            pattern = requirement.get("pattern","")  # the pattern to match. Must be present unless turtle
+
+            if typ[0] == "c":
+                test_string = code
+            elif typ[0] == "f":
+                # get file contents from filename
+                filename = requirement.get("filename")
+                if not filename:
+                    return js.Object.fromEntries(to_js({"outcome": False, "err": "Missing filename in test case", "ins": expected_input}))
+                try:
+                    with open(filename, "r") as f:
+                        test_string = f.read()
+                except FileNotFoundError:
+                    return js.Object.fromEntries(to_js({"outcome": False, "err": "File not found", "ins": expected_input}))
+                except Exception as e:
+                    return js.Object.fromEntries(to_js({"outcome": False, "err": "Unknown error reading file", "ins": expected_input}))
+            elif typ[0] == "s":
+                # get string from evaluating a code statement - find a neater/more efficient way to do this?
+                if "statement" not in requirement:
+                    return js.Object.fromEntries(to_js({"outcome": False, "err": "Missing statement in test case", "ins": expected_input}))
+                try:
+                    test_string = str(eval(requirement.get("statement"), global_vars))
+                except Exception as e:
+                    return js.Object.fromEntries(to_js({"outcome": False, "err": "Error evaluating test-case statement", "ins": expected_input}))
+            elif typ[0] == "t":
+                # evaluate turtle canvas result comparison with code from filename
+                if "filename" not in requirement:
+                    return js.Object.fromEntries(to_js({"outcome": False, "err": "Missing turtle solution filename in test case", "ins": expected_input}))
+                try:
+                    # the filename has been replaced with the soln code
+                    screen_dump_user = run_turtle_cmd({"action":"dump", "value":""})
+                    run_turtle_cmd({"action":"virtual", "value":True})
+                    exec(requirement.get("filename"), global_vars)
+                    screen_dump_soln = run_turtle_cmd({"action":"dump", "value":""})  
+                    run_turtle_cmd({"action":"virtual", "value":False})
```

Comment:
as discussed, we could apply the virtual mode in TS when user hits the submit button (and disable when we get a result). THis simplifies the Python code and also speeds up the user turtle canvas generation.